### PR TITLE
sched-simple: switch discovery to resource.acquire interface

### PR DESCRIPTION
**Pull request by grondo**
This PR updates sched-simple to acquire resources and track their state using the new `resource.acquire` protocol.

sched-simple can now track up/down ranks and exclude down resources from consideration for scheduling.

A simple `flux-resource` frontend is added to allow "drain/undrain" of ranks (using `resource` module), and a very rough cut of a `sched.status` interface allows `flux resource list` to query the current state of resources according to the scheduler.

The current protocol for the latter will definitely need to change, probably even before this gets merged. Definitely still a WIP, but posting now will hopefully generate some discussion.

E.g.
```console
grondo@asp:~/git/flux-core.git$ src/cmd/flux start -s 4
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource list
16/16 cores on 4/4 ranks available
Available: rank[0-3]/core[0-3]
Allocated: 
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini submit sleep 300
358461997056
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource list
15/16 cores on 4/4 ranks available
Available: rank0/core[1-3] rank[1-3]/core[0-3]
Allocated: rank0/core0
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource drain 0
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource list
12/16 cores on 3/4 ranks available
Available: rank[1-3]/core[0-3]
Allocated: rank0/core0
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini submit sleep 300
684913065984
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource list
11/16 cores on 3/4 ranks available
Available: rank1/core[1-3] rank[2-3]/core[0-3]
Allocated: rank0/core0 rank1/core0
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource undrain 0
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource list
14/16 cores on 4/4 ranks available
Available: rank[0-1]/core[1-3] rank[2-3]/core[0-3]
Allocated: rank[0-1]/core0
```

### CODE REVIEW

**Comment by garlick** [comment]
On the `flux resource info` query - I like where you are going with the scheduler's response, which includes full resource objects for `allocated`, and `all` so that the tool can display a finer granularity than execution targets.

The `down` object operates at the execution target level currently, and if provided by the resource module, could be subdivided into "drained" and "offline".

I'm not sure whether it makes sense to have the resource module provide part of the info used in listing.  As an experiment, I added a tthrow-away `query` method to the resource module in 56d8320c6a8b5a75a0d2b3c032872b096b00ef21.  The response looks like this (example 4 brokers, 2 drained):
```json
{
  "resources": {
    "0-3": {
      "Package": 1,
      "Core": 2,
      "PU": 2,
      "cpuset": "0-1"
    }
  },
  "drain": "2-3",
  "offline": ""
}
```
Hmm.  In the course of doing this I realized:
* the above resource object may not be the same as the scheduler's because the exclusion set may have changed
* we need a way to list drained nodes with reason(s)
* maybe `flux resource list` is already giving the appropriate level of detail for _users_, and what I seem to be angling towards is an _admin_ tool

Sorry if I`m getting us off track!  I suspect you have deeper thoughts on this subject, having spent some time pondering Rv1.

**Comment by grondo** [comment]
Yeah, I experimented a bit here with the interface for retrieving the *scheduler* concept of resource availability. (even though a large part of this will need to be redone after we settle on a new resource set (Rv2) format).

The best I could get so far is that scheduler responds to a query with the following resource set objects:

 * *all* - all resources the scheduler is currently tracking
 * *down* - resources the scheduler knows are down, not available
 * *allocated* - the set of all allocated resources.

From these 3 sets, other sets like the set of *up* resources and *available* resources can be computed.  Then if we have a Python module for ingesting and manipulating *R* objects, we can build nice front end tools from that (an experiment for this format is sitting in `flux-resource.py` now). E.g. there can be multiple "display" methods for *R* from the simple (aggregate counts of resources) to complex (`pstree` like display of resource hiearchy)

Another approach to consider would be to add the ability to annotate resources in the new _R_ format with resource states. Then the scheduler can reply with a single object, annotated with up/down/allocated states, and frontend tools can use that interface for query/display of resource sets.

> maybe flux resource list is already giving the appropriate level of detail for users, and what I seem to be angling towards is an admin tool

This is a good insight. Ideally it would be good if the most common queries could be satisfied by the `resource` module, since this would take load off the scheduler. However, I think you're right that users would be most interested in how many resources are available vs utilized, and the resource module isn't able to provide that.

Maybe the scheduler should also have a lighter weight RPC that replies with summary statistics, which might satisfy the majority of queries.

> Sorry if I`m getting us off track! I suspect you have deeper thoughts on this subject, having spent some time pondering Rv1.

No, this is good! I'm really open to anything now, the code here is largely an experiment. About the only insight I've had is that communicating resource state via _R_ is going to be convenient, if we also have (at least) a Python class that can ingest, query, and display _R_.

**Comment by grondo** [comment]
> I'm not sure whether it makes sense to have the resource module provide part of the info used in listing. As an experiment, I added a tthrow-away query method to the resource module in 56d8320. The response looks like this (example 4 brokers, 2 drained):

What does `offline` signify in this response?

**Comment by garlick** [comment]
> What does offline signify in this response?

"down" from a monitoring point of view, e.g. broker hasn't checked in.  I was calling this "down" but then we used that term to include down+drain+(newly-)excluded in the acquire response so thought I would avoid using the same word to mean two different things, although arguably it might be better the other way round (available as a broader category than up).

> Also, if the resource.query response contained a serialized R instead of by_rank object, we could potentially combine the resource.query and sched.resource-status responses, and the "down" object in the scheduler's response could be dropped. Not sure if that is a net win at this point or not.

Yep.  I suppose anything factoring of common code out of multiple schedulers into the single resource module is a win.

**Comment by grondo** [comment]
> "down" from a monitoring point of view, e.g. broker hasn't checked in. I was calling this "down" but then we used that term to include down+drain+(newly-)excluded in the acquire response so thought I would avoid using the same word to mean two different things, although arguably it might be better the other way round (available as a broader category than up).

Ah, good thought. It will be good to get our terminology straight from the beginning.
Though to a scheduler, whether a node is "down" due to being drained or "offline" is somewhat irrelevant, I would personally find calling administratively drained nodes "down" confusing (actually a resource set could be both down/offline _and_ drained).
Also, in the scheduler I was calling "available" resources those available for jobs, i.e. not available as in "online and not drained". I could see how this could get confusing...

> Yep. I suppose factoring common code out of multiple schedulers into the single resource module is a win.

Though all schedulers will need to be able to create a resource set _R_ from their internal representation anyway. So I'm not sure much refactoring would come of it.

**Comment by grondo** [comment]
Ok, I've removed the WIP tag on this PR as I think it is ready for a real review.

The `sched.status` RPC was renamed `sched.resource-status` to be more explicit and to allow future addition of a `sched.status` RPC with perhaps more scheduler rather than resource specific stats.

The sched.status RPC is expected (currently by the `flux resource list` command) to return 3 resource sets in _R_ format:
  * "all" - set of all resources known to the scheduler
  * "down" - set of down resources
  * "allocated" - set of all allocated resources

From the above sets `flux resource` is able to compute
  * "free" set of free resources
  * "up" set of all resources that are not down

`flux resource list` and the `sched.resource-status` are meant to be prototypes or stand-ins until we have a new Rv2 format, at which point they may be augmented or replaced. Consider them proof of concept for now, and thus they do not have extensive testing which would just need to be replaced later. (Only basic functionality testing.)

`flux resource list` itself uses the `flux.util.OutputFormat` class to "format" each resource set, giving a familiar interface for `flux-jobs` users. The currently available list of format fields includes `state`, `nnodes`, `ncores`, `ngpus`, `ranks`, and `rlist` (the sched-simple flat compressed list of cores). Eventually this could be extended to allow arbitrary resource counts and display formats other than `rlist`.

```
usage: flux-resource list [-h] [-v] [-o FORMAT] [-s STATE,...] [-n]

optional arguments:
  -h, --help              show this help message and exit
  -v, --verbose           Include lists of allocated/free resources
  -o, --format=FORMAT     Specify output format using Python's string format
                          syntax
  -s, --states=STATE,...  Output resources in given states
  -n, --no-header         Suppress header output

```

The default `flux resource list` output shows "free", "allocated", and "down" resource sets:
```
$ flux resource list
     STATE NNODES   NCORES    NGPUS
      free      3       12        0
 allocated      0        0        0
      down      1        4        0
```

"up" or "all" resources are not shown by default but could be passed to the `--state` option. With `--verbose` the actual cores in each set are listed in a compressed format:
```
$ flux resource list --state=all,up,down,free,allocated -v
     STATE NNODES   NCORES    NGPUS LIST
       all      4       16        0 rank[0-3]/core[0-3]
        up      3       12        0 rank[1-3]/core[0-3]
      down      1        4        0 rank0/core[0-3]
      free      3       10        0 rank[1-2]/core[1-3] rank3/core[0-3]
 allocated      2        2        0 rank[1-2]/core0
```

File: src/cmd/Makefile.am [hunk 1]
```diff
@@ -68,7 +68,8 @@ dist_fluxcmd_SCRIPTS = \
 	flux-cron \
```
Comment (garlick) [inline src/cmd/Makefile.am:68]: Commit message spelling "implementaition"

File: scripts/.pylintrc [hunk 2]
```diff
@@ -91,7 +91,7 @@ logging-modules=logging
 bad-functions=map,filter,apply,input,file
 
 # Good variable names which should always be accepted, separated by a comma
-good-names=i,j,k,ex,Run,_
+good-names=i,j,k,x,y,z,ex,Run,_
```
Comment (garlick) [inline scripts/.pylintrc:94]: Nice job fixing the linter instead of just renaming variables!

File: src/cmd/Makefile.am (hunk 1 above)
Comment (grondo) [reply src/cmd/Makefile.am:68]: Thanks, fixed and force-pushed.

**Comment by garlick** [comment]
FWIW, the python looks pretty tight to me.  If you feel like more review is unnecessary, just set MWP.  It would be good to have this merged.

**Comment by grondo** [comment]
> FWIW, the python looks pretty tight to me. If you feel like more review is unnecessary, just set MWP. It would be good to have this merged.

The majority of the new Python was added in the *experimental* `Rv1` and `SchedResourceList` classes. These will likely need a bit of work after Rv2 is added anyway, so maybe no point in a thorough review?

My only other comment is that I was forced to add some idset parsing code to the script and it would be nice to instead have a `flux.idset` class that wraps the C library directly, but I wasn't sure how exactly to do that at this point.

**Comment by garlick** [comment]
An idset class in python would be pretty nice.  I wondering wrapping the C is the answer or if it would make sense to build something natively that is compatible?

Anyway, I'm good with skipping the more detailed review if you want to go ahead with MWP.

**Comment by grondo** [comment]
>  I wondering wrapping the C is the answer or if it would make sense to build something natively that is compatible?

Hm, yeah, maybe idset could subclass Python's `set` and just be implemented as a `set` underneath with an added `encode()` method and a constructor that implements the `decode()`? 

I guess I was leaning towrad wrapping libidset.so since so much work went into making the idset small in memory, and expanding large consecutive sets of integers into actual Python `set` would be pretty heavyweight comparatively.  (especially if many of these sets are used in an eventual ResourceSet object)

**Comment by grondo** [comment]
Oh, I forgot, a minor change will be needed in the `flux-sched` testsuite before this PR can be merged. I'll submit that shortly.

**Comment by grondo** [comment]
flux-framework/flux-sched#661 will need to be merged before this can go in.

**Comment by grondo** [comment]
Restarted gh actions since flux-framework/flux-sched#661 was merged.

**Comment by grondo** [comment]
Thanks!

**Pull request merged**